src/Nethermind/Nethermind.Era1/EraImporter.cs

```diff
@@ -0,0 +1,289 @@
+        if (nextEraStart <= to)
+        {
+            for (; blockNumber < nextEraStart; blockNumber++)
+            {
+                await ImportBlock(blockNumber);
+            }
+        }
+
+        // Earlier part can be parallelized
+        long partitionSize = _maxEra1Size;
+        if (from + partitionSize < suggestFromBlock)
+        {
+            ConcurrentQueue<long> partitionStartBlocks = new ConcurrentQueue<long>();
+            for (; blockNumber + partitionSize < suggestFromBlock && blockNumber + partitionSize < to; blockNumber += partitionSize)
+            {
+                partitionStartBlocks.Enqueue(blockNumber);
+            }
+
+            Task[] importTasks = Enumerable.Range(0, 8).Select((_) =>
+            {
+                return Task.Run(async () =>
+                {
+                    while (partitionStartBlocks.TryDequeue(out long partitionStartBlock))
+                    {
+                        for (long i = 0; i < partitionSize; i++)
+                        {
+                            await ImportBlock(i + partitionStartBlock);
+                        }
+                    }
+                });
+            }).ToArray();
+
+            await Task.WhenAll(importTasks);
+        }
+
+        for (; blockNumber <= to; blockNumber++)
+        {
+            await ImportBlock(blockNumber);
+        }
+        elapsed = DateTime.Now.Subtract(lastProgress);
+        LogImportProgress(DateTime.Now.Subtract(startTime), blocksProcessedAtLastLog, elapsed, blocksProcessed, totalblocks);
+
+        if (_logger.IsInfo) _logger.Info($"Finished history import from {from} to {to}");
+
+        async Task ImportBlock(long blockNumber)
+        {
+            cancellation.ThrowIfCancellationRequested();
+
+            (Block? b, TxReceipt[]? r) = await eraStore.FindBlockAndReceipts(blockNumber, cancellation: cancellation);
+            if (b is null)
+            {
+                throw new EraImportException($"Unable to find block info for block {blockNumber}");
+            }
+            if (r is null)
+            {
+                throw new EraImportException($"Unable to find receipt for block {blockNumber}");
+            }
+            if (b.Number != blockNumber)
+            {
+                throw new EraImportException($"Unexpected block number. Expected {blockNumber}. Got {b.Number}");
+            }
+
+            if (b.IsGenesis)
+            {
+                return;
+            }
+
+            if (b.IsBodyMissing)
+            {
+                throw new EraImportException($"Unexpected block without a body found for block number {blockNumber}. Archive might be corrupted.");
+            }
+
+            if (suggestFromBlock <= b.Number)
+            {
+                await pacer.WaitForQueue(b.Number, cancellation);
+                await SuggestAndProcessBlock(b);
+            }
+            else
+                InsertBlockAndReceipts(b, r);
```

Comment:
Do we need to decode blocks and receipts from RLP just to insert it here again and code it again (maybe we do)

Reply:
Yes.